### PR TITLE
ARROW-9288: [C++][Dataset] Fix PartitioningFactory with dictionary encoding for HivePartioning

### PR DESCRIPTION
**Pull request by jorisvandenbossche**

### CODE REVIEW

File: cpp/src/arrow/dataset/partition.cc [hunk 1]
```diff
@@ -646,15 +657,26 @@ class HivePartitioningFactory : public PartitioningFactory {
...
-    return impl.Finish(&dictionaries_);
+    auto schema_result = impl.Finish(&dictionaries_);
+    field_names_ = impl.FieldNames();
+    return schema_result;
   }
 
   Result<std::shared_ptr<Partitioning>> Finish(
       const std::shared_ptr<Schema>& schema) const override {
-    return std::shared_ptr<Partitioning>(new HivePartitioning(schema, dictionaries_));
+    for (FieldRef ref : field_names_) {
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/dataset/partition.cc:667]: I should probably guard here against the case that `field_names_` was not yet updated (if `Finish` is called without `Inspect` being called), with empty vector?

**Review by jorisvandenbossche — COMMENTED** [review]
The approach here is to also determine `field_names_` in HivePartitioningFactory after inspecting (for DirectoryPartitioningFactory, those field names are passed to the constructor). So that we can then trim the schema and have the `dictionaries` match the order of the schema.

However, thinking of it now: there might still be a problem if the user specified the full dataset schema so no inspection happens .. So we might need to think of a better solution.

(I should also add some C++ tests)

File: cpp/src/arrow/dataset/partition.cc (hunk 1 above)
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/partition.cc:667]: Absolutely, the first line of this method should just call

```
auto field_names = FieldNames();
```

and replace occurrences of the private member.

Comment (jorisvandenbossche) [reply cpp/src/arrow/dataset/partition.cc:667]: There is no `FieldNames()` method on the PartitioningFactory (only the `impl` has one, but that is not accessible here; that's the reason I added the `field_names_` private member to store those)

File: cpp/src/arrow/dataset/partition.cc [hunk 2]
```diff
@@ -646,15 +657,26 @@ class HivePartitioningFactory : public PartitioningFactory {
...
+      // ensure all of field_names_ are present in schema
+      RETURN_NOT_OK(ref.FindOne(*schema).status());
+    }
+
+    // drop fields which aren't in field_names_
+    auto out_schema = SchemaFromColumnNames(schema, field_names_);
+
+    return std::make_shared<HivePartitioning>(std::move(out_schema), dictionaries_);
```
Comment (bkietz) [inline cpp/src/arrow/dataset/partition.cc:675]: The check against field_names_ is only relevant if dictionaries_ is non-empty, which can only occur if Inspect has been called (and `field_names_` has therefore been initialized)
```suggestion
    if (dictionaries_.empty()) {
      return std::make_shared<HivePartitioning>(schema, dictionaries_);
    } else {
      for (FieldRef ref : field_names_) {
        // ensure all of field_names_ are present in schema
        RETURN_NOT_OK(ref.FindOne(*schema).status());
      }

      // drop fields which aren't in field_names_
      auto out_schema = SchemaFromColumnNames(schema, field_names_);

      return std::make_shared<HivePartitioning>(std::move(out_schema), dictionaries_);
    }
```

File: cpp/src/arrow/dataset/partition.cc [hunk 3]
```diff
@@ -646,15 +657,26 @@ class HivePartitioningFactory : public PartitioningFactory {
       }
     }
 
-    return impl.Finish(&dictionaries_);
+    auto schema_result = impl.Finish(&dictionaries_);
+    field_names_ = impl.FieldNames();
+    return schema_result;
```
Comment (bkietz) [inline cpp/src/arrow/dataset/partition.cc:662]: Finish doesn't mutate `name_to_index_` and that is the only data member accessed by `FieldNames()`. I don't see why Finish needs to be called first
```suggestion
    field_names_ = impl.FieldNames();
    return impl.Finish(&dictionaries_);
```

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/dataset/partition.cc (hunk 2 above)
Comment (jorisvandenbossche) [reply cpp/src/arrow/dataset/partition.cc]: Thanks, that's a nice way by checking `dictionaries_` to ensure that `field_name_` is set or not

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**